### ISSUE TITLE
Performance issue with org-roam

### ISSUE THREAD

**Issue by hungpham2511** [org-roam/org-roam#834]
### Description

I have enjoyed using org-roam thus far and been using that quite extensively. Though I can't help noticing that the mode becomes sluggish very quickly when there are a few hundred notes. This is getting a bit annoying and I end up turning off org-roam more and more often.

It seems to me that org-roam is trying to rebuild the cache all the time at saving hook, and enforce auto-save after a few seconds. 
I wonder if there is any way to strip the hooks to a bare minimum or any algorithmic improvement to improve performance?

- Org-roam commit: https://github.com/jethrokuan/org-roam/commit/7ab67436a7517ee6be35c32b96c3c6d697181603

**Comment by hungpham2511** [comment]
I might not be fully aware of the internal implementation. However, I find it quite strange that this update step should be expensive at all? Assuming one 1 file changes at save, we only need to update the db of the entries that are referenced to in this note, if there are any. In the current implementation, I notice that the delay occurs even when there are very few backlinks.

**Comment by jethrokuan** [comment]
I have about 500 notes and I'm not seeing any noticeable lag at all. The db updates are incremental, and they are implemented how you describe it.

Can you use `M-x profiler-start` and go along your business (editing notes, saving them) and `M-x profiler-stop`, and then `M-x profiler-report`?

That would help me see where the delays are.

**Comment by isolet** [comment]
I have a similar performance issue that `org-roam` calls lots of `file-truename` and causes slowness, though the original callee is `org-roam-bibtex`. Below is the profile of a single call of `save-buffer` of my bib file. The file contains only 1000 items and I have only 200 org file in my `org-roam-directory`. It took 5 to 10 seconds to save my bib file.


```
- command-execute                                                1846  99%
 - call-interactively                                            1846  99%
  - funcall-interactively                                        1846  99%
   - file-notify-handle-event                                    1741  93%
    - file-notify--callback-kqueue                               1741  93%
     - file-notify--handle-event                                 1741  93%
      - file-notify--call-handler                                1741  93%
       - #<compiled 0x1ff20ab7bbe1>                              1741  93%
        - bibtex-completion-candidates                           1741  93%
         - bibtex-completion-parse-bibliography                  1699  91%
          - bibtex-completion-prepare-entry                      1658  88%
           - mapcar                                              1372  73%
            - #<compiled 0x1ff20ab8272d>                         1372  73%
             - orb-find-note-file                                1243  66%
              - let*                                             1243  66%
               - org-roam--get-ref-path-completions               1243  66%
                - org-roam-db-query                              1243  66%
                 - org-roam-db                                    934  50%
                  - org-roam-db--get-connection                   934  50%
                   - file-truename                                934  50%
                    - file-truename                               792  42%
                     - file-truename                              574  30%
                      - file-truename                             404  21%
                       - file-truename                            262  14%
                        - file-truename                           158   8%
                           file-truename                           35   1%
                 - apply                                          309  16%
                  - emacsql                                       309  16%
                   - apply                                        309  16%
                    - #<compiled 0x1ff20b934e51>                  309  16%
                     - emacsql-send-message                       158   8%
                      - apply                                     158   8%
                       - #<compiled 0x1ff20a3348e9>                158   8%
                        - apply                                   158   8%
                         - #<compiled 0x1ff20b93e98d>                158   8%
                          - emacsql-sqlite3-run-dot-command                 54   2%
                             mapconcat                              1   0%
                     - emacsql-wait                               149   7%
                      - apply                                     149   7%
                       - #<compiled 0x1ff20b932e35>                149   7%
                        - emacsql-waiting-p                         1   0%
                         - apply                                    1   0%
                          - #<compiled 0x1ff20b93e971>                  1   0%
                           - rx-to-string                           1   0%
                              apply                                 1   0%
                     - apply                                        2   0%
                      - emacsql-compile                             2   0%
                       - apply                                      1   0%
                        - emacsql-format                            1   0%
                           apply                                    1   0%
                         emacsql-prepare                            1   0%
             - bibtex-completion-find-note-multiple-files                103   5%
                f-directory?                                       76   4%
                f-file?                                            27   1%
             - bibtex-completion-find-note-one-file                 26   1%
                f-file?                                            26   1%
           - bibtex-completion-find-pdf                           286  15%
            - bibtex-completion-find-pdf-in-library                286  15%
             - -first                                             283  15%
                f-file?                                           283  15%
             - bibtex-completion-get-value                          3   0%
                replace-regexp-in-string                            2   0%
          - parsebib-read-entry                                    40   2%
           - parsebib--find-bibtex-field                           35   1%
            - parsebib--parse-value                                29   1%
             - parsebib--expand-strings                            23   1%
              - mapcar                                             22   1%
               - #<compiled 0x1ff20eeb3ad5>                        22   1%
                  replace-regexp-in-string                         15   0%
                  match-string                                      1   0%
             - parsebib--match-delim-forward                        3   0%
              - parsebib--match-quote-forward                       3   0%
                 forward-sexp                                       1   0%
           - parsebib--match-paren-forward                          1   0%
            - parsebib--match-brace-forward                         1   0%
               forward-sexp                                         1   0%
         - bibtex-completion-resolve-crossrefs                     22   1%
          - bibtex-completion-make-candidate                       16   0%
           - bibtex-completion-clean-string                        15   0%
              replace-regexp-in-string                             11   0%
             -map                                                   1   0%
          - bibtex-completion-make-entry-hash                       6   0%
           - bibtex-completion-get-value                            6   0%
            - s-collapse-whitespace                                 5   0%
             - replace-regexp-in-string                             5   0%
                apply                                               1   0%
              replace-regexp-in-string                              1   0%
         - bibtex-completion-parse-strings                          3   0%
          - parsebib-find-next-item                                 3   0%
             parsebib--looking-at-goto-end                          1   0%
         - secure-hash                                              2   0%
          - select-safe-coding-system                               2   0%
           - find-coding-systems-region                             2   0%
            - sort-coding-systems                                   1   0%
             - sort                                                 1   0%
              - #<compiled 0x1ff20ae6556d>                          1   0%
               - #<compiled 0x1ff20ae6555d>                         1   0%
                  coding-system-category                            1   0%
           f-file?                                                  1   0%
   - counsel-M-x                                                   97   5%
    - ivy-read                                                     74   3%
     - read-from-minibuffer                                        29   1%
      - ivy--queue-exhibit                                          9   0%
       - ivy--exhibit                                               9   0%
        - ivy--update-minibuffer                                    9   0%
         - ivy--format                                              7   0%
          - mapcar                                                  6   0%
           - ivy-rich--counsel-M-x-transformer                      6   0%
            - ivy-rich-format                                       6   0%
             - mapconcat                                            6   0%
              - #<compiled 0x1ff20a8142e1>                          6   0%
               - ivy-rich-format-column                             6   0%
                  ivy-rich-counsel-function-docstring                  5   0%
                  counsel-M-x-transformer                           1   0%
          - ivy--wnd-cands-to-str                                   1   0%
           - ivy-format-function-line                               1   0%
              ivy--format-function-generic                          1   0%
         - ivy--filter                                              2   0%
          - ivy--re-filter                                          2   0%
           - cl-remove                                              2   0%
            - apply                                                 2   0%
             - cl-delete                                            2   0%
                #<compiled 0x1ff20a809169>                          1   0%
      - redisplay_internal (C function)                             6   0%
       - company-posframe-window-change                             5   0%
        - company-posframe-hide                                     5   0%
         - posframe-hide                                            5   0%
            posframe--make-frame-invisible                          5   0%
       - #<compiled 0x43f82bcb>                                     1   0%
        - apply                                                     1   0%
         - redisplay--pre-redisplay-functions                       1   0%
            run-hook-with-args                                      1   0%
     - ivy--reset-state                                             3   0%
      - ivy--filter                                                 2   0%
       - ivy--re-filter                                             1   0%
          cl-remove                                                 1   0%
         ivy--recompute-index                                       1   0%
     - ivy-call                                                     2   0%
      - counsel-M-x-action                                          2   0%
       - command-execute                                            2   0%
        - call-interactively                                        2   0%
         - funcall-interactively                                    2   0%
            profiler-report                                         2   0%
    - counsel--M-x-externs                                         23   1%
     - smex-detect-new-commands                                     5   0%
      - mapatoms                                                    5   0%
         #<compiled 0x1ff207502471>                                 3   0%
   - save-buffer                                                    8   0%
    - basic-save-buffer                                             7   0%
     - basic-save-buffer-1                                          6   0%
      - basic-save-buffer-2                                         6   0%
       - write-region                                               2   0%
        - select-safe-coding-system                                 2   0%
         - find-auto-coding                                         1   0%
            sgml-html-meta-auto-coding-function                     1   0%
           find-coding-systems-region                               1   0%
     - vc-after-save                                                1   0%
      - vc-state-refresh                                            1   0%
       - vc-call-backend                                            1   0%
        - apply                                                     1   0%
         - vc-git-state                                             1   0%
          - apply                                                   1   0%
           - vc-git--run-command-string                             1   0%
            - apply                                                 1   0%
             - vc-git--out-ok                                       1   0%
              - apply                                               1   0%
               - vc-git--call                                       1   0%
                - apply                                             1   0%
                 - process-file                                     1   0%
                    apply                                           1   0%
    - message                                                       1   0%
     - apply                                                        1   0%
        message-filter                                              1   0%
+ redisplay_internal (C function)                                  11   0%
+ timer-event-handler                                               6   0%
+ ...                                                               0   0%
```

**Comment by jethrokuan** [comment]
@zaeph would probably have a better idea about this. org-roam doesn't do much work if the save-buffer isn't an org-roam file (i.e. org file, and in the org-roam directory). So saving `.bib` files shouldn't have this kind of slowdown.

**Comment by zaeph** [comment]
I believe that, by 'bib file', @isolet meant his Org-mode files where all his bibliographic notes are.  Neither Org-roam nor ORB are optimised to handle large Org-mode files because we need to parse them entirely every time.  That being said, I'm not sure why `file-truename` is run that many times.

Would you be able to send me your notes so that I may run some tests on my end?  Feel free to expunge whatever data you want with a regex, as long as you keep the structure and the links intact.

**Comment by jethrokuan** [comment]
Using @nobiot 's test org files, here's my profiler report:

```
          - org-roam-db-build-cache                    18,734,498,553  99%
           + org-roam--extract-links                    6,448,491,611  34%
           + org-roam-db-query                          3,693,151,753  19%
           + org-roam--extract-titles                   2,981,193,735  15%
           + org-roam--extract-tags                     1,536,131,808   8%
           + org-roam--extract-ref                      1,483,872,058   7%
           + org-roam--extract-headlines                1,467,978,194   7%
           + org-roam-db--clear-file                      406,911,880   2%
```

Expanding the tree, I see that `org-element-parse-buffer` dominates the time spent. I'm going to try once again to reduce the number of necessary calls to that.

**Comment by isolet** [comment]
@zaepzeIt's a large bibtex file.

**Comment by hungpham2511** [comment]
I had a closer look into my issue and apparently deft and company-mode are the actual culprit that slow things down for small org files. Basically deft has an auto-save hook that keeps trigger org-roam hooks. I am not so sure about company-mode though, but the latency probably comes from it trying to look through all the files to provide suggestions.

Regarding big org-files (~10k lines), I do experience some performance issues during saving. Is it possible to rebuild links from the diff of the current buffer content and the file on disk? That would definitely help improving performance. Integration with a version control system like git, basically using its efficient glob diffing mechanism would be nice.

**Comment by zaeph** [comment]
The idea you've mentioned about working with diffs is interesting, but we're going to need to prototype something.  We're brainstorming with @jethrokuan to see if it's reasonably doable.

**Comment by ddxgz** [comment]
I'm also experiencing the same performance issue, besides big org files, a small file with about 500 lines but many links will also have a 1 sec lag.

**Comment by nobiot** [comment]
@jethrokuan, moving from #859 ([your request](https://github.com/org-roam/org-roam/issues/859#issuecomment-647304879) to benchmark on Windows). 

Benchmark results below. Took about 4.5 hours (overnight).
[A copy of profile ](https://drive.google.com/drive/folders/1PuRCMlayev_hWB8onGULalAA2f3v-iqK)shared here (hope this works).

I will try my Linux (WSL) on the same machine tonight to see if it's really Windows...

```
*** 1000 org files from Jethro
#+begin_src emacs-lisp
(setq org-roam-directory "~/test-org-files-master")
(benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
;: Elapsed time: 16417.908832s (8.850320s in 1 GCs)
```
In addition, for comparison I did this based on my 100 org files (not from your 1000 files)
```
*** 100 org files
#+begin_src emacs-lisp
(setq org-roam-directory "~/temp/10000-org-files-linked/100-org-linked-files")
(benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
;: Elapsed time: 190.369353s (0.230377s in 1 GCs)
```

**Comment by jethrokuan** [comment]
That's a crazy long time, it's 50s on Linux for 1000 files in comparison, will have to look into what's taking up all that time.

**Comment by nobiot** [comment]
@jethrokuan ,

There is something about change to transaction:
- Back to commit `eca0727` (release): Org-roam v1.2.0 (#802 
- Windows 

Results below.
Slower than your Linux machine, but 253 seconds is OK compared to 4.5 hours.

```
#+begin_src emacs-lisp
(setq org-roam-directory "~/test-org-files-master-before-transaction")
(benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
;: Elapsed time: 253.065005s (6.878271s in 72 GCs)
```


I'm also running my Linux on the same machine -- literally still running. I expect it to be some hours to go.
I'll come back here for update. 

Is there any flag / setting for SQLite outside Emacs? Your CPU architecture (mine is Intel)? Or is it the way I run this (compared to your way)?

**Comment by jethrokuan** [comment]
@nobiot there were 2 changes made in that commit (transactions, and GC threshold changing). Can you help me confirm that it's the change to transactions that severely affects performance time by setting the following:

```
(setq org-roam-db-gc-threshold gc-cons-threshold)
```

From what I see, I don't think GC would have caused this issue, because little time is spent during GC.

Also, when you run your benchmarks, can you run `(org-roam-db-build-cache t)`? The additional parameter `t` instructs org-roam to trash the old db before rebuilding, simulating a full rebuild. if you change directories and run a rebuild, it will have to do some processing based on content in the old db (checking if there is a file in there already processed, deleting files that are gone etc.).

**Comment by nobiot** [comment]
I'll do try gc-cons-threashold and see. I

just to be sure, though, not sure `(org-roam-db-build-cache t)` would change the outcome (?)
For every benchmark run I do, I manually delete the db file, or unzip from the master to start from scratch. So... I thought effectively I was doing the same thing?

Do you still want me to try this? -- I might just do this with 100 notes to begin with to see if there is any difference. 

For the latest one with heading support, the very first db build created .org-id-locations (with nil) in my emacs.d.  I am not touching that in this benchmark.

**Comment by jethrokuan** [comment]
@nobiot yup if you manually delete the db file, then the outcome should be the same.

**Comment by nobiot** [comment]
!!!! See the result below. 

It's GC then?  Take this as my first attempt.
With such a dramatic change, I am doubting myself. Let me try this a couple of times to make sure my setups, etc. are all correct.
But I cannot hide my excitement!! (down from 4.5 hours).

Do you see anything glaringly wrong here?

```
*** 1000 org files from Jethro [2020-06-25] v1.2 79c75ac with GC (Windows)
#+begin_src emacs-lisp
;; manually deleted the org-roam.db
(setq org-roam-db-gc-threshold gc-cons-threshold)
(setq org-roam-directory "~/test-org-files-master")
(benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
: Elapsed time: 159.885425s (4.207748s in 70 GCs)
```

Compare this with 2 days ago without `gc-cons-threshold`:
```
*** 1000 org files from Jethro
#+begin_src emacs-lisp
(setq org-roam-directory "~/test-org-files-master")
(benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
;: Elapsed time: 16417.908832s (8.850320s in 1 GCs)
```

**Comment by nobiot** [comment]
@jethrokuan 
What do you think? 
I am happy that we can have instant improvement, but am puzzled by the behavior.

- `(setq org-roam-db-gc-threshold gc-cons-threshold)` dramatically improves the performance: around 2-6 minutes on Doom
- Doom does *something* that makes the process faster (compared to vanilla)
- Increasing threshold higher than Doom's default 16MB to 64MB slows down to 15 minutes
- Vanilla (+OR and ORB config; threshold 80K) is significantly slower than Doom: around 30 minutes

Detailed notes here:
https://gist.github.com/nobiot/a28cd6405eeb075cc07c7428be86555e

**Comment by TimQuelch** [comment]
Doom uses GCMH to manage its gc scheduling. That could be contributing

**Comment by jethrokuan** [comment]
wow okay, so it looks like GCing often is better for Windows, but seems to be worse for Linux, at least on my machine. I'm not sure if Doom should've made a difference: GCMH seems to set a higher threshold for GC, but other than that there should be no noticeable difference.

So at least we can conclude that the GCs are taking up the time, rather than the transactions. Now what should be the way forward? Do we reset the default value for `org-roam-db-gc-threshold` to something lower that works across all machines, and have users tweak it when they want to achieve better performance?

**Comment by nobiot** [comment]
@jethrokuan , 
My sense is to set the default value of `org-roam-db-gc-threshold` to what Doom does (16MB) or set it to `gc-cons-threshold`, across all platforms, for now at least.

It's an interesting one. We have not seen how macOS behaves...

Your machine might be a special one -- is your CPU ARM (not sure if it is relevant)?  See my Ubuntu experiment below. It's the same dramatic difference from 240 minutes to 2 minutes. It's via WSL, so I believe it should behave in the native Linux way on the same machine. 

@TimQuelch ,  I want to also try GCMH, or at least the way Doom sets up different thresholds (high, and etc., I think). If you are interested to see how your machine behaves (a Linux distro?), I'd love to see the outcome...

```
My Ubuntu Linux via WSL

org-roam-20200624.530
[2020-06-24]                                                                                                          #+begin_src emacs-lisp
  (require 'org-roam)
  (setq org-roam-directory "~/test-org-files-master")
  (benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
: Elapsed time: 14430.126326s (25.395115s in 1 GCs)

[2020-06-25]
#+begin_src emacs-lisp
;; manually deleted the org-roam.db
  (require 'org-roam)
  (setq org-roam-db-gc-threshold gc-cons-threshold)
  (setq org-roam-directory "~/test-org-files-master")
  (benchmark 1 '(org-roam-db-build-cache))
#+end_src

#+RESULTS:
: Elapsed time: 137.406451s (22.518389s in 1033 GCs)

```

**Comment by jethrokuan** [comment]
I don't think my PC is special, because this is also the runtime of the CI, and it matches my observations on my PC.

Nonetheless if it is causing issues, I will set it to `gc-cons-threshold` by default and add some docs on tweaking it, until we can figure out what's causing the slowdown.

I have some ideas that will mitigate the need for much GC (Not parsing the AST and just doing forward regex searches, but will need to see how that pans out)

**Comment by nobiot** [comment]
I like the idea of forward regex search.

**Comment by charignon** [comment]
To add some data to this conversation. I benchmarked org-roam on my set of org-roam files (~700 of them), it generally takes 500ms to build the DB incrementally and 7s from scratch. I found that most of the time is spent parsing org mode files like you suggest here. 

I tried building a prototype with regex search using https://github.com/borkdude/babashka: https://github.com/charignon/org-roam which brings down the time to build from scratch to under 200ms in my case so I can always build from scratch. On the 1000 test org files it also takes around 200ms to rebuild from scratch. So I think that we are going in the right direction by not using the org parser.
